File: ruby/problems/n-queens.rb
class Solver
  attr_accessor :n, :board, :cols, :diag1, :diag2, :res

  def initialize(n)
    @n = n
    @board = Array.new(n) { Array.new(n, '.') }
    @cols = Array.new(10, false)
    @diag1 = Array.new(20, false)
    @diag2 = Array.new(20, false)
    @res = []
  end

  def solve(i)
    if i == n
      add_solution
      return
    end

    @n.times do |j|
      next unless !cols[j] && !diag1[i + j] && !diag2[n - 1 + j - i]

      board[i][j] = 'Q'
      cols[j] = diag1[i + j] = diag2[n - 1 + j - i] = true

      solve(i + 1)

      board[i][j] = '.'
      cols[j] = diag1[i + j] = diag2[n - 1 + j - i] = false
    end
  end

  def add_solution
    cur = board.map(&:join)
    res.push(cur)
  end
end

#  param {Integer} n
# @return {String[][]}
def solve_n_queens(n)
  s = Solver.new(n)
  s.solve(0)
  s.res
end

File: python/problems/n-queens.py
class Solution:
  def solveNQueens(self, n: int) -> List[List[str]]:
    board = [['.'] * n for _ in range(n)]
    cols = [False] * 10
    diag1 = [False] * 20
    diag2 = [False] * 20
    res = []

    def add_solution():
      res.append([''.join(x) for x in board])

    def solve(i: int):
      if i == n:
        add_solution()
        return

      for j in range(n):
        if not cols[j] and not diag1[i + j] and not diag2[n - 1 + j - i]:
          board[i][j] = 'Q'
          cols[j] = diag1[i + j] = diag2[n - 1 + j - i] = True

          solve(i + 1)

          board[i][j] = '.'
          cols[j] = diag1[i + j] = diag2[n - 1 + j - i] = False

    solve(0)
    return res
